### PR TITLE
NBS: Fragmentation tool using new estimate of locality

### PR DESCRIPTION
**Pull request by cmasone-attic**
The new version of this tool now estimates the locality of a DB
written using the "grandchild" strategy implemented by
types.ValueStore. It does do by dividing each level of the graph
up into groups that are roughly the size of the branching factor
of that level, and then calculating how many physical reads are
needed to read each group.

In the case of perfect locality, each group could be read in a
single physical read, so that's what the tool uses as its estimate
of the optimal case.

Toward #2968

### CODE REVIEW

File: go/nbs/table_set.go [hunk 1]
```diff
@@ -86,16 +86,17 @@ func (ts tableSet) getMany(reqs []getRecord, foundChunks chan *chunks.Chunk, wg
 }
 
 func (ts tableSet) calcReads(reqs []getRecord, blockSize uint64) (reads int, split, remaining bool) {
-	f := func(css chunkSources) (reads int, split, remaining bool) {
+	f := func(css chunkSources) (int, bool, bool) {
```
Comment (cmasone-attic) [inline go/nbs/table_set.go]: This was busted, so I fixed it as a part of this work

**Comment by cmasone-attic** [comment]
PTAL. New sync strategy in a separate patch. Seems to do WAY better than the current version, at least by the metric provided by this tool :-)

File: go/nbs/frag/main.go [hunk 2]
```diff
@@ -63,72 +62,85 @@ func main() {
...
-						for _, r := range children {
-							refs <- r
-						}
-					}()
+		// Iterate all the Values at the current level of the graph IN ORDER (as specified in |current|) and gather up their embedded refs. We'll build two different lists of hash.Hashes during this process:
+		// 1) An ordered list of ALL the children of the current level.
+		// 2) An ordered list of the child nodes that contain refs to chunks we haven't yet visited. This *excludes* already-visted nodes and nodes without children.
+		// We'll use 1) to get an estimate of how good the locality is among the children of the current level, and then 2) to descend to the next level of the graph.
+		orderedChildren := hash.HashSlice{}
+		nextLevel := hash.HashSlice{} // TODO: re-use |current| to save allocations?
```
Comment (rafael-atticlabs) [inline go/nbs/frag/main.go]: nah.

File: go/nbs/table_reader.go [hunk 3]
```diff
@@ -389,6 +389,7 @@ func canReadAhead(fRec offsetRec, fLength uint32, readStart, readEnd, blockSize
 		return readEnd, true
 	}
 
+	// Is this right? This seems like it means that we're willing to keep reading ahead as long as the next chunk is < blockSize away from the end of the last chunk. That may be what we want, but then it's not really a "block size". It's more like...the amount of overhead we're willing to tolerate per chunk read.
```
Comment (rafael-atticlabs) [inline go/nbs/table_reader.go]: real comment

**Pull request merged**